src/compiler/ir/README.md

```diff
@@ -0,0 +1,162 @@
+should provide storage for a pointer to the closure's body, the list of cowns
+and the captures.
+
+The closure's body will accept as many parameters as there are cowns and
+captures specified. When invoking the body, each cown argument should be
+transformed into a reference to its contents, before being passed as an
+argument. The captures should be passed as is.
+
+## `ReadFieldStmt`
+
+Read a field from an object. The field is identified by its name. The general
+case is similar to method calls: the field's selector is used to look up the
+field's offset in the object's vtable.
+
+## `WriteFieldStmt`
+
+Write a field into an object. The procedure for locating the offset of the
+field is identical to a field read.
+
+When writing a reference counted value to an object in a region that is
+equipped with a RememberedSet, the value should be inserted into the RS.
+Otherwise, the reference count should be incremented.
+
+## `IntegerLiteralStmt`, `StringLiteralStmt`, `UnitStmt`
+
+These create values from a literal. The support for strings is very special
+cased for demonstration purposes.
+
+## `MatchBindStmt`
+
+This is an unchecked cast, used at the top of the target of a match statement.
+
+## `ViewStmt`
+
+If the argument to the statement is an owned value, the statement produces an
+unowned reference to the same object. Otherwise this is equivalent to a copy.
+
+## `EndScopeStmt`
+
+This statement is inserted whenever variables go out of scope. This may not be
+the end of a BasicBlock, if a block expression is used for example.
+
+This is where a value's "release" code should be executed. This could be
+`rt::Region::release`, `rt::Cown::release` or `rt::Immutable::release`, for
+owned, cown and immutable values, respectively. Unowned values do not need to
+be released.
+
+## `OverwriteStmt`
+
+This is used after a local variable assignement, on the old SSA value bound to
+this statement. The old value should be released, as in an `EndScopeStmt`. For
+example, the code `var x = new A; x = new B; ...`, would be lowered to the
+following IR:
+
+```
+0 <- new A
+1 <- new B
+overwrite(0)
+// ...
+end-scope(1)
+```
+
+The local `x` is originally assigned the SSA variable `0`, and later `1`. After
+the assignement, the `overwrite(0)` statement encodes the fact that the old
+value needs to be released.
+
+Note that an SSA value that has been overwritten is never passed to an
+"end-scope" statement. Only `x`'s new value, `1`, goes out of scope.
+
+# Terminators
+
+## Branch Terminator
+## If Terminator
+## Match Terminator
+
+# Builtins
+
+A number of language features don't need special syntax, but are instead
+exposed as builtin methods, defined in `stdlib/builtin.verona`. This includes
+arithmetic operations and some region operations such as freeze and trace.
+
+The language features they represent isn't even expressible in the IR. Instead,
+`codegen/builtins.cc` provides hardcoded lowerings into bytecode.
```

Comment:
IIUC, we'll still need to carry those as calls to functions, ex. `U32.add(x, y)` in MLIR, and then have a special lowering to LLVM that picks the codegen alternative and either call the intrinsic or lower to a native operation, right?

Reply:
So in my opinion, `x + y` is always lowered as a method call (`x.add(y)`), all the way through to LLVM. We also generate an add method which is lowered using an intrinsic / native operation. Inlining is responsible for replacing the method call with the native operation, not the lowering.

If we find out that inlining is too unreliable / extra work, we can do the lowering to native ops at a later point. But I don't see the point in doing that yet.

Reply:
I think Sylvan said this was an issue with Pony, that the LLVM inliner was unreliable and that's why they lowered to intrinsic/op directly into LLVM. 

One thing that could be a problem is if we lower calls with more than just `call` (like lowered vtable access). In that case, `x.add(y)` would actually be a sequence of instructions which culminates in a dynamic call (call on the offset of the vtable), and LLVM would have no idea what function is actually being called.

The simple case of a static call (like `U32.add(x,y)`) would be converted into a call to a mangled function, available in the library and lowered in the same LLVM IR, and then the inliner could spot that and inline. But we'd still need to convert from `x.add(y)` to `TypeOfX.add(x, y)` before lowering to LLVM.

Makes sense?

Reply:
The issue is that this will not always be a statically dispatched call. If you consider the slightly simpler `-x` expression, it could be that `x` has type `Neg`, where `Neg` is defined as :
```
interface Neg {
  neg(self: imm): Neg & imm
}
```

Now the expression `-x` (which is really `x.neg()`) cannot be dispatched to a specific method but instead does the usual vtable lookup and dynamic dispatch stuff. But I32 has this `neg` method, therefore there must actually exist a vtable for `I32`, which has a pointer to the body of the method. That body cannot be implemented in Verona code and must be some compiler magic eg. in the bytecode backend this is the `builtins.cc` file that directly emits bytecode instructions that have no equivalent in the IR.

Since the code path for treating arithmetic operations as functions, and lowering of the function bodies, must exist for this purpose, we can always use it and 1) hope LLVM inlines it in the concrete arithmetic type case 2) do something better ourselves at a later point if this is a problem.

Reply:
Right, I consider this to be a language definition problem. 

If we guarantee native type arithmetic will always be lowered to machine instructions on registers, we need to build interface/library/lowering in a way that allows us to achieve this without complex hacks in the compiler.

Relying on the compiler matching functions by name on "known" exceptions can lead to horrible and poorly maintained code. 

Perhaps we could add a `native` keyword that means "this is implemented outside of the language" or something like this, so that at least we know what must and what must not be done in that way and develop the library around that concept.

Reply:
Yes that’s what the `builtin.verona` file already does. A number of methods in there are tagged with a `builtin` keyword.